fix: normalize inputAmount by stripping leading zeros after first non…

nikaaru opened the pull request:
## 📝 Summary

- Allows users to type raw-zero values like `0000`, `0.`, or `0.0000` without immediate formatting
- As soon as a non-zero digit is entered, strips leading zeros and normalizes input (e.g. `00001` → `1`, `0000.01` → `0.01`)
- Enforces a minimum acceptable value of `0.00001` — input below this (e.g. `0.00000`) is blocked

### 🔍 Changes

- Added four regex constants in **UPPER_SNAKE_CASE**:
  - `ALL_ZERO_REGEX` – detects pure-zero strings (e.g., `0000`, `0.000`)
  - `LEADING_ZEROS_REGEX` – strips leading zeros only when followed by a digit
  - `LEADING_DOT_REGEX` – normalizes inputs like `.5` → `0.5`
  - `ALLOWABLE_ZERO_INPUT_REGEX` – allows partial raw-zero input like `0.`, `0.000`

- Updated `setInputAmount`:
  - Allows raw-zero phase inputs (via `ALL_ZERO_REGEX` + `ALLOWABLE_ZERO_INPUT_REGEX`)
  - Sanitizes only after the first meaningful digit using `LEADING_ZEROS_REGEX` and `LEADING_DOT_REGEX`
  - Blocks values below `0.00001` by checking parsed numeric value

### 🛠️ How to Test
1. Run the app and enter each scenario from the table below.  
2. Ensure the input behaves as expected and normalization occurs only when appropriate.

### ✅ Some Test Scenarios
| Input        | Expected Output | Notes                                  |
|--------------|-----------------|----------------------------------------|
| `00000`      | `00000`         | raw-zero, allowed                      |
| `0.`         | `0.`            | raw-zero + dot, allowed                |
| `0.0000`     | `0.0000`        | up to 4 decimal zeros allowed          |
| `00001`      | `1`             | sanitized                              |
| `0000.01`    | `0.01`          | sanitized + normalized                 |
| `.5`         | `0.5`           | normalized                             |
| `0.00001`    | `0.00001`       | meets minimum, allowed                 |



## 🔄 Updates (since last review)
- Extracted `sanitize/validate` logic into `sanitizer.ts` & `validation.ts`
- Added `sanitizer.test.ts` & `validation.test.ts` with edge‑case coverage
- Removed constants from `constants.ts` (`ALL_ZERO_REGEX`, `LEADING_ZEROS_REGEX`, `LEADING_DOT_REGEX`, `ALLOWABLE_ZERO_INPUT_REGEX`) — now internalized in `sanitizer.ts` and `validation.ts.`
- Removed "minimum value" validation (no longer blocks `0.00000`)
- Refactored and extracted shared logic into new utility modules:
   `sanitizer.ts`: includes `formatThousandsWithCommas`, `replaceSpacesWithDash`
   `validation.ts`: includes `isValidCurrencyFormat`, `isNumeric`, `isColorKeyOverridden`
- Added unit tests for all extracted utility functions
- Added sanitizeInputAmount method to normalize zero values (e.g., `0000` → `0`) on input **blur**
- Sync **URL** search params with `quoteStore`.

#### 🧪 Verification Steps

- Review new files `sanitizer.ts` and `validation.ts` for correct function extraction.
- Run a `yarn test` and confirm that all tests in `sanitizer.test.ts` and `validation.test.ts` pass.

yeager-eren reviewed (commented):
Why you didn't use [zod](https://zod.dev/?id=numbers)‍‍?

yeager-eren commented:
1. Query string should be validated and fixed on first page load. [e.g. ](http://rango-widget-iframe-3m27y1ljv-rango-exchange.vercel.app/?fromBlockchain=ETH&fromToken=ETH&toBlockchain=ETH&toToken=USDT--0xdac17f958d2ee523a2206206994597c13d831ec7&fromAmount=000.000)
2. Blur event should be used also to update and fix user's input. e.g. type 0000 and click outside the input (blur).

nikaaru commented:
> Why you didn't use [zod](https://zod.dev/?id=numbers)‍‍?

I'm not using Zod here because the project doesn’t include it before, adding Zod now would bloat our bundle without user‑visible error messages as required.
And also, it wasn't mentioned in the task description either.

nikaaru commented:
> 1. Query string should be validated and fixed on first page load. [e.g. ](http://rango-widget-iframe-3m27y1ljv-rango-exchange.vercel.app/?fromBlockchain=ETH&fromToken=ETH&toBlockchain=ETH&toToken=USDT--0xdac17f958d2ee523a2206206994597c13d831ec7&fromAmount=000.000)
> 2. Blur event should be used also to update and fix user's input. e.g. type 0000 and click outside the input (blur).

1. The query string was also validated, but it wasn't truncated properly, so I fixed it.
2. Blur's behavior is 1inch different, and also this was not mentioned in the task.

yeager-eren commented on widget/embedded/src/utils/validation.ts:
```diff
@@ -0,0 +1,39 @@
+ * @param input - string to test, e.g. "0", "10.25"
+ * @returns true for valid money formats, false otherwise
+ */
+export function isValidCurrencyFormat(input: string) {
```
nit: i don't know change the name or not, currency format is a good name, but not all the currencies have two minor unit.
Take a look at [ISO 4217](https://www.iso.org/iso-4217-currency-codes.html) and [this list](https://www.six-group.com/dam/download/financial-information/data-center/iso-currrency/lists/list-one.xls).

nikaaru replied:
I added dynamic decimal support to the `isValidCurrencyFormat` function.
Also, I updated a unit test.

yeager-eren commented on widget/embedded/src/utils/sanitizers.test.ts:
```diff
@@ -0,0 +1,87 @@
+      expect(ensureLeadingZeroForDecimal('000.000')).toBe('000.000');
+    });
+
+    test('should not alter other inputs', () => {
```
how about `0.1a`?

nikaaru replied:
Should reject the non-numeric input.

I added some rejected tests for this function:

```ts
    test('should rejects non-numeric and malformed strings', () => {
      expect(isValidCurrencyFormat('')).toBe(false);
      expect(isValidCurrencyFormat('.5')).toBe(false);
      expect(isValidCurrencyFormat('1.2.3')).toBe(false);
      expect(isValidCurrencyFormat('abc')).toBe(false);
      expect(isValidCurrencyFormat('0.1a')).toBe(false);
    });
    ```

yeager-eren reviewed (approved):
neat implementation. thanks.

nikaaru commented on widget/embedded/src/utils/sanitizers.test.ts:
```diff
@@ -0,0 +1,87 @@
+
+    test('should convert tabs and newlines into dashes', () => {
+      const words = faker.lorem.words(WORD_COUNT).split(' ');
+      console.log(replaceSpacesWithDash(words.join('\t')));
```
Removed

yeager-eren reviewed (commented):
I resolved my comments. you can move it to test.
